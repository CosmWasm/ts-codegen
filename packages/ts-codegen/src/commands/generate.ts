import { prompt } from '../utils/prompt';
import codegen from '../index';
import { TSBuilderOptions } from '../builder';

export default async (argv) => {
    const questions = [
        {
            _: true,
            type: 'path',
            name: 'schema',
            message: 'which directory contains the the Rust contracts?',
            default: './schema'
        },
        {
            _: true,
            type: 'path',
            name: 'out',
            message: 'where is the output directory?',
            default: './ts'
        },
        {
            _: true,
            type: 'string',
            name: 'name',
            message: 'contract name?'
        },
        {
            type: 'checkbox',
            name: 'plugin',
            message: 'which plugins?',
            choices: [
                'client',
                'recoil',
                'react-query',
                'message-composer',
                'msg-builder',
                'abstract-app'
            ]
        },
        {
            type: 'confirm',
            name: 'bundle',
            message: 'enable bundle?',
            default: true
        }
    ];

    if (argv.typesOnly) {
        argv.plugin = 'types';
    }

    let { schema, out, name, plugin, bundle } = await prompt(questions, argv);
    if (!Array.isArray(plugin)) plugin = [plugin];

    ///////// REACT QUERY
    const questions2 = [];
    if (plugin.includes('react-query')) {
        [].push.apply(questions2, [
            {
                type: 'confirm',
                name: 'optionalClient',
                message: 'optionalClient?',
                default: false
            },
            {
                type: 'list',
                name: 'version',
                message: 'which react-query version?',
                default: 'v3',
                choices: ['v3', 'v4']
            },
            {
                type: 'confirm',
                name: 'queryKeys',
                message: 'queryKeys?',
                default: false
            },
        ])
    };
    const { optionalClient, version, queryKeys } = await prompt(questions2, argv);
    const questions3 = [];
    if (version === 'v4') {
        [].push.apply(questions3, [
            // currently we only support v4 for useMutation
            {
                type: 'confirm',
                name: 'mutations',
                message: 'Generate useMutation hooks?',
                default: false
            }

        ])
    };
    const { mutations } = await prompt(questions3, argv);

    const queryFactoryQuestions  = [];
    if (queryKeys) {
      [].push.apply(queryFactoryQuestions, [
        // Only can use queryFactory if queryKeys is enabled
        {
          type: 'confirm',
          name: 'queryFactory',
          message: 'queryFactory?',
          default: false
        }
      ])
    };
    const { queryFactory } = await prompt(queryFactoryQuestions, argv);
    ///////// END REACT QUERY

    ///////// BUNDLE
    const questions4 = [];
    if (bundle) {
        [].push.apply(questions4, [
            {
                type: 'string',
                name: 'bundleFile',
                message: 'bundleFile?',
                default: 'index.ts'
            },
            {
                type: 'string',
                name: 'bundleScope',
                message: 'bundleScope?',
                default: 'contracts'
            }
        ])
    };
    const { bundleFile, bundleScope } = await prompt(questions4, argv);
    ///////// END BUNDLE

    const options: TSBuilderOptions = {
        types: {
            enabled: true
        },
        client: {
            enabled:
                plugin.includes('client') ||
                plugin.includes('recoil') ||
              // react-query either uses client or abstract-app
              (plugin.includes('react-query') && !plugin.includes('abstract-app'))
        },
        reactQuery: {
            enabled: plugin.includes('react-query'),
            optionalClient,
            queryKeys,
            version,
            mutations,
            queryFactory
        },
        recoil: {
            enabled: plugin.includes('recoil'),
        },
        messageComposer: {
            enabled: plugin.includes('message-composer')
        },
        msgBuilder: {
          enabled: plugin.includes('msg-builder') || plugin.includes('abstract-app')
        },
        bundle: {
            enabled: bundle,
            scope: bundleScope,
            bundleFile
        },
        abstractApp: {
          enabled: plugin.includes('abstract-app'),
        }
    };

    await codegen({
        contracts: [
            {
                name,
                dir: schema
            }
        ],
        outPath: out,
        options
    })
};
