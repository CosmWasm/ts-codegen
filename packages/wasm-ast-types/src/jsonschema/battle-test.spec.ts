import generate from '@babel/generator';
import { sync as glob } from 'glob';
import { readFileSync } from 'fs';
import {
    RenderContext,
    processTypes,
} from './jsonschema-types'
import cases from 'jest-in-case';

const minter = glob(__dirname + '/../../../../__fixtures__/minter/*.json').map(file => {
    return { name: file.split('__fixtures__')[1], content: JSON.parse(readFileSync(file, 'utf-8')) }
});

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

const context: RenderContext = {
    options: {
        optionalArrays: true
    }
}

cases('minter', opts => {
    processTypes(
        context,
        opts.content
    ).map(ast => {
        expectCode(ast);
    })
}, minter);

it('processTypes', () => {

});