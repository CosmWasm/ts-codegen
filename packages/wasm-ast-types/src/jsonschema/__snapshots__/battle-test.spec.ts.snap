// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minter /Users/pyramation/code/cosmwasm/ts-codegen/__fixtures__/minter/execute_msg.json 1`] = `
"export type ExecuteMsg = {
  mint: object;
} | {
  set_whitelist: {
    whitelist: string;
  };
} | {
  update_start_time: Timestamp;
} | {
  update_per_address_limit: {
    per_address_limit: number;
  };
} | {
  mint_to: {
    recipient: string;
  };
} | {
  mint_for: {
    recipient: string;
    token_id: number;
  };
} | {
  withdraw: object;
};"
`;

exports[`minter /Users/pyramation/code/cosmwasm/ts-codegen/__fixtures__/minter/execute_msg.json 2`] = `"export type Timestamp = Uint64;"`;

exports[`minter /Users/pyramation/code/cosmwasm/ts-codegen/__fixtures__/minter/execute_msg.json 3`] = `"export type Uint64 = string;"`;

exports[`minter /Users/pyramation/code/cosmwasm/ts-codegen/__fixtures__/minter/query_msg.json 1`] = `
"export type QueryMsg = {
  config: object;
} | {
  mintable_num_tokens: object;
} | {
  start_time: object;
} | {
  mint_price: object;
} | {
  mint_count: {
    address: string;
  };
};"
`;

exports[`minter /minter/execute_msg.json 1`] = `
"export type ExecuteMsg = {
  mint: object;
} | {
  set_whitelist: {
    whitelist: string;
  };
} | {
  update_start_time: Timestamp;
} | {
  update_per_address_limit: {
    per_address_limit: number;
  };
} | {
  mint_to: {
    recipient: string;
  };
} | {
  mint_for: {
    recipient: string;
    token_id: number;
  };
} | {
  withdraw: object;
};"
`;

exports[`minter /minter/execute_msg.json 2`] = `"export type Timestamp = Uint64;"`;

exports[`minter /minter/execute_msg.json 3`] = `"export type Uint64 = string;"`;

exports[`minter /minter/query_msg.json 1`] = `
"export type QueryMsg = {
  config: object;
} | {
  mintable_num_tokens: object;
} | {
  start_time: object;
} | {
  mint_price: object;
} | {
  mint_count: {
    address: string;
  };
};"
`;

exports[`minter case: 3 1`] = `
"export type ExecuteMsg = {
  mint: object;
} | {
  set_whitelist: {
    whitelist: string;
  };
} | {
  update_start_time: Timestamp;
} | {
  update_per_address_limit: {
    per_address_limit: number;
  };
} | {
  mint_to: {
    recipient: string;
  };
} | {
  mint_for: {
    recipient: string;
    token_id: number;
  };
} | {
  withdraw: object;
};"
`;

exports[`minter case: 3 2`] = `"export type Timestamp = Uint64;"`;

exports[`minter case: 3 3`] = `"export type Uint64 = string;"`;

exports[`minter case: 5 1`] = `
"export type QueryMsg = {
  config: object;
} | {
  mintable_num_tokens: object;
} | {
  start_time: object;
} | {
  mint_price: object;
} | {
  mint_count: {
    address: string;
  };
};"
`;

exports[`processTypes 1`] = `
"export type ExecuteMsg = {
  mint: object;
} | {
  set_whitelist: {
    whitelist: string;
  };
} | {
  update_start_time: Timestamp;
} | {
  update_per_address_limit: {
    per_address_limit: number;
  };
} | {
  mint_to: {
    recipient: string;
  };
} | {
  mint_for: {
    recipient: string;
    token_id: number;
  };
} | {
  withdraw: object;
};"
`;

exports[`processTypes 2`] = `"export type Timestamp = Uint64;"`;

exports[`processTypes 3`] = `"export type Uint64 = string;"`;
